### ISSUE TITLE
Enhancement: Custom error messages

### ISSUE THREAD

**Issue by aubergine10** [DavidAnson/markdownlint#23]
## Rules

Currently rules have a static error message, it would be useful if they could generate context-sensitive errors, for example:

```js
errors.push({
  lineNumber: whatever,
  error: `The '${headingText}' heading should be at level ${expectedLevel}`
});
```

## Styles

An optional top-level error message that gets output upon the first error for a given batch of documents.

```js
options.config._guidance = 'Please refer to the style guide for more information on the errors below: http://whatever.com/styleguide'
```

This would be particularly useful in CI builds of open source projects to help guide contributors to the style guide or templates, etc.

**Comment by DavidAnson** [comment]
Makes total sense. Will think about this more and look into it soon! Much appreciated. :)

**Comment by aubergine10** [comment]
On the subject of custom rule errors, it would be useful if configurable rules stated the config in their error message. For example, I just got these CI errors on some doc updates for eslint:

```
docs/rules/sort-imports.md: 164: MD004 Unordered list style
docs/rules/sort-imports.md: 37: MD007 Unordered list indentation
```

I had to search build scripts to work out what params had been sent to the linter, for example how many spaces should be used for list indentation? It would save time if errors looked like this:

```
docs/rules/sort-imports.md: 164: MD004 Unordered lists should use {asterisk} style
docs/rules/sort-imports.md: 37: MD007 Unordered list indentation should be {2} spaces
```

That will enable me, and others, to see at a glance what needs changing to fix an error.

**Comment by DavidAnson** [comment]
This ended up being a bigger task than I expected, but I think it's a significant improvement, thanks for suggesting it!

PS - I did not implement a guidance message because it seems like configuration/process that will be handled differently by various projects (documentation, output message, code comment, etc.).
